### ISSUE TITLE
Documentation does not explain how to select a style

### ISSUE THREAD

**Issue by ysalmon** [tweh/menukeys#46]
Hello,
I have just discovered your package ; it is quite nice !

However, the PDF documentation seems to lack explanation about how to select one the style presented in Section 4.2. Indeed, this section details everything about styles, but not how to produce the different examples provided in Subsection 4.2.1.

This is extremely frustrating !

I resorted to looking into the source code of the manual and it seems that using ```\renewmenumacro{\keys}{shadowedangularkeys}``` for example works. Yet I am not sure if this is the correct way to change the style for displaying keys, as this command is only mentioned in relation with menu styles, which is also consistent with its name.

I think the documentation would be improved by adding a new Subsection just before current Subsection 4.2.1 with title "Selecting a style".

**Comment by tweh** [comment]
Hi,

you’re right, this isn't well documented and if I find some time I’ll take a look to improve the manual.

And your guess is right

    \renewmenumacro{\keys}{<style>}

is the way to go. An ”menu style” ist meant in a broad meaning here. Actually the package doesn’t differentiate between menu, keys, paths or whatever. It is just a sequence of items with some styling for each item (depending ion the position in the sequence).

**Comment by thymaro** [comment]
I had the exact same problem. I was skimming the manual to see what I was doing wrong when selecting the style. I had to comment out the style specification in my preamble, because I couldn't get them to work and I'm not very fluent in tex-package-sourcecode. I am glad to having found this thread, but it should really be in the documentation. 😜

**Comment by ysalmon** [comment]
I went to report this, only to remember that I had already !

The interface to specify a style, besides being undocumented, is very counter-intuitive.

One tends to try

``\usepackage[<style>]{menukeys}``

and

``\usepackage[style=<style>]{menukeys}``

**Comment by Skillmon** [comment]
@ysalmon I could provide such options, it shouldn't be too difficult. Does the following interface sound good to you?

For each of the default macros provide a key accepting the different styles:

- option `menus` accepting the values `default`, `rounded`, and `angular`
- option `keys` accepting the values `rounded`, `shadowedrounded`, `angular`, `shadowedangular` and `typewriter`
- option `paths` accepting the values `default`, `withfolder`, `withblackfolder`, `hyphenate`, `hyphenatewithfolder`, `hyphenatewithblackfolder`

**Comment by ysalmon** [comment]
I am not so sure because this would only work for the default available styles.

In fact, everything would make more sense if the ``\renewmenumacro`` command was called ``\setmenustyle`` or something like that.

This is a naming convention and documentation problem rather than a functionality problem.

More generally, the documentation is bizarre : it tells us how to define new styles (a feature for advanced users), it explains the source code of the package (which is only interesting to people who might want to fork it), but it does not tells what a user discovering the package should type in order to get the sample outputs of section 4.2.

This is true for setting the style, but also pertaining to the ``+`` symbol between keys.

When I set ``\renewmenumacro{\keys}{roundedkeys}`` and give ``\keys{\ctrl+c}``, I get a single rectangle with ``ctrl+c`` within it, not two rectangles separated with a ``+``. And this is confusing, because obviously ``ctrl+c`` is two keys, not one, and because section 4.2.1 shows output with multiples rectangles separated by ``+``. So I wonder *how* this outpout was obtained.

**Comment by tweh** [comment]
Hi guys,

I’d say that this in fact is documented in section *4.4.2 Defining or changing menu macros*. So maybe the documentation is bad but complete ;-)

I’d rathe not use package options for changing the styles as – like ysalmon said – this would only work for the predefined ones.

Regarding your concern about the naming convention: Since the `\(re)newmenumacro` (re)defines a new macro it would be wrong in my eyes to rename ist to simple something like `\changestyle`.

However as this problem just seems to be to find the solution I’d suggest, we add something like the following in the documentation right before presenting the styles:

----

**4.2.1 Predefined styles**

The package offers some predefined style to be used with existing and new menu macros. To creat a new one with a style you can add 
```latex
\newmenumacro{\mymenus}{roundmenus}
```
to the preamble. To change the styl of an existing macro you can use
```latex
\newmenumacro{\mymenu}{roundmenus}
```
instead. Pleas note that this might also changes the separation character. See section 4.4.2 for more details on how to (re)define macros.

**Comment by tweh** [comment]
> This is true for setting the style, but also pertaining to the + symbol between keys.

When I set \renewmenumacro{\keys}{roundedkeys} and give \keys{\ctrl+c}, I get a single rectangle with ctrl+c within it, not two rectangles separated with a +. And this is confusing, because obviously ctrl+c is two keys, not one, and because section 4.2.1 shows output with multiples rectangles separated by +. So I wonder how this output was obtained.

This is also explained in section 4.4. However it could be worth to discuss the order of the section, but again, I won't say that there is something undocumented. And pleas don’t complain about the fact that the implementation is also documented. You don’t have to read this part, but it’s common sense that this should be documented ;-)

**Comment by tweh** [comment]
PS: The thin with the separator is also explained in 4.1 so at the very beginning and before talking about the styles. See page 5 and footnote 4. So maybe you could expand on how you think this ist not documented …

**Comment by ysalmon** [comment]
I am not complaining ;-) It is always good to document code — but that is another story, because the user of some software or package is not concerned with the code, only with the interface.

Thank you for pointing out the relevant parts of the documentation. I think I finally understand what had been bothering me.

The notion of style you are using is not the same as the one in eg. TikZ or beamer (which is odd in part because if I understand correctly this package relies on tikz). In TikZ, changing the style of an element does not involve defining or redefining a macro : the predefined macros fetch the needed styles. This is also the case with eg. CSS : you use it to style preexisting elements, you do not redefine elements to give the a style.

This is why I would never have thought about looking at a part of a documentation pertaining to (re)defining macros in order to choose the style of the output : section 4.4.2 is about defining or changing menu macro, but when trying to apply a style, this is not what I have in mind.

While this way of doing things makes sense once you get it (I think it is somehow similar to how theorem styles work in most packages), it does not seem consistent with the way color themes are managed (which is more beamer/tikz-like) ; it also has the bizarre consequence about the separator : even though the separator for ``\keys`` is ``+`` by default, I need to restate it explicitly if I ``\renewmacro`` it, because for that command, the default separator is ``,``, and this even if I was only concerned with changing the appearance of the keys.

I hope I managed to convey more clearly what I previously found unclear ;-)

**Comment by Skillmon** [comment]
@ysalmon of course a `.style` in Ti*k*Z redefines a macro, that's just not transparent to the user...

@tweh of course this is documented (which is why I only addressed the suggested interface request of @ysalmon), but I should've pointed it out in the first place... And I like your suggestion of adding a brief introduction to `\(re)newmenumacro` before showing the different styles, and I think with this the documentation will be much clearer and the confusion as described by @ysalmon would be avoided in the future.

**Comment by ysalmon** [comment]
Yes of course ; but the fact that setting a style involves changing macros internally is not a concern for the user : semantically, this is not the same concept at all to him/her. In the same fashion that some other features of some packages involve changing catcodes internally : the intent of the user when invoking such features is not to change catcodes, and they should not be concerned with that — it is a shortcoming of TeX that they sometimes have to.

**Comment by tweh** [comment]
@ysalmon thanks for the clarification. Now I got what you mean :-)

For me as a non-native “style” means the appearance of something and thus this wording. I see your pint that there is a differnce in TikZ and CSS where object and their style are somehow separated. But for `menukeys` I see no advantage in this. Otherwise it would require a greater change in the interface, to something like

```latex
\newmenumacro{\menu}{
   style = rounded,
   separator = >,
  color-theme = black,
  …
}
```

However even with this change I won’t say that `\(re)newmenumacro` is a wrong wording. Since wie deal with clerical TeX macros in this case other than in TikZ where you can’t redefine the object like a `node` itself but only its style. I hope to be clear …

**Comment by p10n** [comment]
I was also struggling, because I saw that the doc mentions how to define a style, but before I do that I would rather test the existing ones, which look nice to me. But that is not even mentioned. The update in the doc seems a little help, but did not make it into my LaTeX distro (though quite recent).

I even asked chatGPT how to select a style, and it did not provide the working answer. I think the current way of selecting a style is really very unusual.

I really think that the suggestion above https://github.com/tweh/menukeys/issues/46#issuecomment-1399267672 is a perfect interface and probably highly appreciated by most users who just want to select one of the default styles without digging deeper into the documentation. I would not care too much about exact wording (CSS etc.) - that is imo beyond this ticket :-)

**Comment by tweh** [comment]
Just as a quick  note: on p. 12 in the manual (v1.6.2) it is explained how to define and change the macros, you can just use the macro and select a style.

```latex
\⟨re⟩newmenumacro{⟨macro⟩}[⟨input sep⟩]{⟨style⟩}.
```

But I agree, that it might be worth to mention it in the styles section as well. However at the moment I have zero spare time unfortunately so overhauling the interface/syntax should not be expected to happen …

**Comment by tweh** [comment]
@Skillmon Could you take a look please? In the repo there is already an addition to the document (commit
c0be7e1946f64ead17d6c32b9e501bb904df240d) but this never made it to CTAN. Do you know why?

**Comment by Skillmon** [comment]
@tweh yeah, will do, but I don't have too much spare time myself these days. Maybe I can spend an hour or so later today (sidenote: GitHub is acting up on me, the browser UI doesn't load at all, I'm sending this comment from the CLI).

**Comment by Skillmon** [comment]
The commit you indicate doesn't contain any wording improvement on the documentation. I just added a `\changes`-entry for a PR I merged earlier because the changes were omitted there.

And 1.6.2 is the latest released version on CTAN (so contains https://github.com/tweh/menukeys/commit/c0be7e1946f64ead17d6c32b9e501bb904df240d). I just forgot to add a tag on the released commit it seems.

**Comment by Skillmon** [comment]
~~So we still need to improve the documentation here as this seems to be a recurring issue for many readers.~~ See below.

**Comment by Skillmon** [comment]
@tweh But adc704e does indeed contain a short explanation at the start of the styles section. No idea why that never made its way to CTAN, sorry. (in my defence: you merged that :P)

**Comment by tweh** [comment]
HI, yes I didn’t want to say it's your fault nor didn't I wanted to assign this test to you. It's just that you made the changes with the `\DeclareVersion` etc. and I was unsure, what that was about. Anyway thanks for submitting! Shall we close this issue now?

**Comment by Skillmon** [comment]
I haven't submitted the update, yet. There are issues with the index (I solved them on my previous laptop simple-mindedly by changing the build environment to TL'21, but on my current laptop I don't reach that far back, and now I can't build the documentation). I'll figure something out, revert some changes I made to the Makefile because of that, and probably do some minor clean up stuff and then release. I think I'll be done with all that on Monday evening (I don't have much time myself, as I previously said).

The `\DeclareVersion` stuff is to allow for LaTeX's built-in package rollback support to work. I don't know whether that's really needed, but since that version possibly changed the spacing of `menukeys` at the end of lines and other edge cases I figured it was a good idea to support it. If you load the package with `\usepackage{menukeys}[=v1.6.1]` you'll get the old behaviour back.

**Comment by Skillmon** [comment]
@tweh what do you think of the following additional change?

<img width="717" height="954" alt="Image" src="https://github.com/user-attachments/assets/b3a9f8d9-54aa-4676-a20f-15a7cda2e8dd" />

<img width="712" height="1051" alt="Image" src="https://github.com/user-attachments/assets/7362cd49-8509-43b8-9c10-401965c0413b" />

**Comment by p10n** [comment]
Thanks for your quick response, and for your effort! Much appreciated.

I think providing directly working examples is the best idea, it makes life so much easier for me as user because I can copy-paste it and I am ready to go, for a start. Thus I like your suggestion of providing the necessary commands right at the spot where the styles are presented.

However, in particular, I do not understand why I should _renew_ a macro called `\test` here?
I guess `\test` would be a new one, that I need to define?

BTW, if I do not specify any style, which one is being used as a default? For menus, for keys, for paths? That would also be useful info that I missed in the documentation.

**Comment by Skillmon** [comment]
@p10n if it's not too much hassle, can you clone this and compile the current master on your local PC? I fixed the examples to show `\newmenumacro\test{...}` on the very first example and use `\renewmenumacro` on later ones. Also, I added some very simple minded syntax highlighting to these examples only.

The reason my screenshots show `\renewmenumacro` is because that's what the examples in the documentation do, so it was only the natural idea here.

**Comment by tweh** [comment]
@Skillmon Yea, that's a nice idea! Thanks!

**Comment by Skillmon** [comment]
I just pushed v1.6.3 containing the documentation improvements to CTAN. Thanks to everybody involved here this will hopefully be no issue in the future anymore.
